src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -4345,32 +4578,101 @@ private uint GetTupleValEscape(ImmutableArray<BoundExpression> elements, uint sc
-                    var rightValEscape = assignment.IsRef
-                        ? GetRefEscape(assignment.Right, scopeOfTheContainingExpression)
-                        : GetValEscape(assignment.Right, scopeOfTheContainingExpression);
+                var exprResult = GetValEscapeOfObjectMemberInitializer(expr, scopeOfTheContainingExpression);
+                result = Math.Max(result, exprResult);
+            }
 
-                    result = Math.Max(result, rightValEscape);
+            return result;
+        }
+
+        private uint GetValEscapeOfObjectMemberInitializer(BoundExpression expr, uint scopeOfTheContainingExpression)
+        {
+            uint result;
+            if (expr.Kind == BoundKind.AssignmentOperator)
+            {
+                var assignment = (BoundAssignmentOperator)expr;
+                var rightEscape = assignment.IsRef
+                    ? GetRefEscape(assignment.Right, scopeOfTheContainingExpression)
+                    : GetValEscape(assignment.Right, scopeOfTheContainingExpression);
 
-                    var left = (BoundObjectInitializerMember)assignment.Left;
-                    result = Math.Max(result, GetValEscape(left.Arguments, scopeOfTheContainingExpression));
+                uint leftEscape;
+                var left = (BoundObjectInitializerMember)assignment.Left;
+                if (left.MemberSymbol is PropertySymbol { IsIndexer: true } property)
+                {
+                    Debug.Assert(left.AccessorKind != AccessorKind.Unknown);
+                    MethodInfo methodInfo = MethodInfo.Create(property, left.AccessorKind);
+                    leftEscape = getIndexerEscape(methodInfo, left, rightEscape);
                 }
                 else
                 {
-                    result = Math.Max(result, GetValEscape(expression, scopeOfTheContainingExpression));
+                    leftEscape = GetValEscape(left.Arguments, scopeOfTheContainingExpression);
+                    leftEscape = Math.Max(leftEscape, rightEscape);
                 }
+
+                result = leftEscape;
+            }
+            else
+            {
+                result = GetValEscape(expr, scopeOfTheContainingExpression);
             }
 
             return result;
+
+            uint getIndexerEscape(
+                in MethodInfo methodInfo,
+                BoundObjectInitializerMember expr,
+                uint rightEscapeScope)
+            {
+                Debug.Assert(methodInfo.Method is not null);
+
+                // If the indexer is readonly then none of the arguments can contribute to 
+                // the receiver escape
+                if (methodInfo.Method.IsEffectivelyReadOnly)
+                {
+                    return CallingMethodScope;
+                }
+
+                var escapeValues = ArrayBuilder<EscapeValue>.GetInstance();
+                GetEscapeValuesForUpdatedRules(
```

Comment:
>GetEscapeValuesForUpdatedRules [](http://example.com/codeflow?start=16&length=30)

We're using updated rules always. Is that correct? (It's not clear whether we get here when compiling with older language versions.)

Reply:
Don't think it functionally matters for this particular case. But this was a mistake. I forgot to come back and update this after I refactored the `GetEsacpeValues` to support old / updated rules.